File: src/algorithms/strings/pattern-matching/naive-pattern-search/__tests__/naive-pattern-search_test.go
package main

import "testing"

func TestNaivePatternSearchPatternAtStart(t *testing.T) {
	if naivePatternSearch("ABCDEF", "ABC") != 0 {
		t.Error("expected 0")
	}
}

func TestNaivePatternSearchPatternInMiddle(t *testing.T) {
	if naivePatternSearch("AABAACAADAABAABA", "AABA") != 0 {
		t.Error("expected 0")
	}
}

func TestNaivePatternSearchPatternAtEnd(t *testing.T) {
	if naivePatternSearch("XYZABC", "ABC") != 3 {
		t.Error("expected 3")
	}
}

func TestNaivePatternSearchPatternNotFound(t *testing.T) {
	if naivePatternSearch("ABCDEFG", "XYZ") != -1 {
		t.Error("expected -1")
	}
}

func TestNaivePatternSearchSingleCharFound(t *testing.T) {
	if naivePatternSearch("HELLO", "L") != 2 {
		t.Error("expected 2")
	}
}

func TestNaivePatternSearchSingleCharNotFound(t *testing.T) {
	if naivePatternSearch("HELLO", "Z") != -1 {
		t.Error("expected -1")
	}
}

func TestNaivePatternSearchEmptyPattern(t *testing.T) {
	if naivePatternSearch("HELLO", "") != 0 {
		t.Error("expected 0 for empty pattern")
	}
}

func TestNaivePatternSearchTextEqualsPattern(t *testing.T) {
	if naivePatternSearch("ABCD", "ABCD") != 0 {
		t.Error("expected 0")
	}
}

func TestNaivePatternSearchPatternLongerThanText(t *testing.T) {
	if naivePatternSearch("AB", "ABCD") != -1 {
		t.Error("expected -1")
	}
}

func TestNaivePatternSearchRepeatedChars(t *testing.T) {
	if naivePatternSearch("AAAAAB", "AAAB") != 2 {
		t.Error("expected 2")
	}
}

func TestNaivePatternSearchWorstCase(t *testing.T) {
	if naivePatternSearch("AAAAAAB", "AAAAB") != 2 {
		t.Error("expected 2")
	}
}

File: src/algorithms/strings/pattern-matching/naive-pattern-search/naive-pattern-search_test.rs
include!("sources/naive-pattern-search.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pattern_at_start() {
        assert_eq!(naive_pattern_search("ABCDEF", "ABC"), 0);
    }

    #[test]
    fn test_pattern_in_middle() {
        assert_eq!(naive_pattern_search("AABAACAADAABAABA", "AABA"), 0);
    }

    #[test]
    fn test_pattern_at_end() {
        assert_eq!(naive_pattern_search("XYZABC", "ABC"), 3);
    }

    #[test]
    fn test_pattern_not_found() {
        assert_eq!(naive_pattern_search("ABCDEFG", "XYZ"), -1);
    }

    #[test]
    fn test_single_char_found() {
        assert_eq!(naive_pattern_search("HELLO", "L"), 2);
    }

    #[test]
    fn test_single_char_not_found() {
        assert_eq!(naive_pattern_search("HELLO", "Z"), -1);
    }

    #[test]
    fn test_empty_pattern() {
        assert_eq!(naive_pattern_search("HELLO", ""), 0);
    }

    #[test]
    fn test_text_equals_pattern() {
        assert_eq!(naive_pattern_search("ABCD", "ABCD"), 0);
    }

    #[test]
    fn test_pattern_longer_than_text() {
        assert_eq!(naive_pattern_search("AB", "ABCD"), -1);
    }

    #[test]
    fn test_repeated_characters() {
        assert_eq!(naive_pattern_search("AAAAAB", "AAAB"), 2);
    }

    #[test]
    fn test_worst_case_repetitive() {
        assert_eq!(naive_pattern_search("AAAAAAB", "AAAAB"), 2);
    }
}
